동민/10주차/10주차_2617.kt

```diff
@@ -0,0 +1,63 @@
+import java.io.BufferedReader
+import java.io.BufferedWriter
+import java.io.InputStreamReader
+import java.io.OutputStreamWriter
+
+var cntLTH = 0 // 본인보다 무거운 구슬을 세는 카운트
+var cntHTL = 0 // 본인보다 가벼운 구슬을 세는 카운트
+
+fun main(args: Array<String>) = with(BufferedReader(InputStreamReader(System.`in`))){
+    val (N, M) = readLine().split(" ").map { it.toInt() }
+
+    var graph = Array(N+1, { IntArray(N+1,{0}) }) // 그래프 생성
+    repeat(M){
+        val (fir, sec) = readLine().split(" ")
+        graph[fir.toInt()][sec.toInt()] = 1 // 본인보다 무거운 구슬을 1로 표기
+        graph[sec.toInt()][fir.toInt()] = 2 // 본인보다 가벼운 구슬을 2로 표기
+    }
+
+    var half = (N+1)/2 // 중간 이상 벗어났는지 확인하기 위한 변수
+    var result = 0 // 결과값
+    for (i in 1 .. N.toInt()){
+        cntLTH = 0 // 초기화
+        cntHTL = 0 // 초기화
+        var visitedLowToHigh = BooleanArray(N+1, {false}) // 방문기록 -> 본인보다 무거운것 확인
+        var visitedHighToLow = BooleanArray(N+1, {false}) // 방문기록 -> 본인보다 가벼운것 확인
+
+        dfsLowToHigh(graph, visitedLowToHigh, i) // DFS 실행 -> 본인보다 무거운 것
+        dfsHighToLow(graph, visitedHighToLow, i) // DFS 실행 -> 본인보다 가벼운 것
+
+        if (cntHTL >= half || cntLTH >= half) // 본인보다 가볍거나 무거운게 절반 이상 있을경우 중간이 절대 될 수 없음
+            result++
+    }
+
+    val bw = BufferedWriter(OutputStreamWriter(System.out))
+
+    bw.write("${result}")
+    bw.flush()
+    bw.close()
+}
+
+// DFS -> 방문하지 않았고, 본인보다 무거운 경우(1) 카운트를 늘리고 DFS 재귀 호출
+fun dfsLowToHigh(graph: Array<IntArray>, visited: BooleanArray, positon: Int){
+    visited[positon] = true
+
+    for (i in 0 until graph[positon].size){
+        if (!visited[i] && graph[positon][i] == 1){
```

Comment:
graph[position][i]의 값이 1과 2로 구분되는것만으로는 명확하게 의미를 이해하기가 어렵습니다.
따라서 이 부분을 조금 더 명확하게 변경해주는게 좋을 것 같습니다.

하나로 dfsHighToLow와 dfsLowToHigh를 합쳤다고 가정하여 적어보겠습니다.
문법이 틀렸을수도 있는데 저렇게 쓰는게 좋을거같네요

```suggestion
        // 상수는 최상단에다가 선언
        const val HEAVY = 1;
        const val LIGHT = 2;
        if (!visited[i] && graph[positon][i] == HEAVY){
```